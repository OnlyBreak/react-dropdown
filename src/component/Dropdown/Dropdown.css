.br-dropdown {
  position: relative;
  display: inline-block;
  color: #48576a;
  font-size: 12px;
}

.br-dropdown-inner {
  position: relative;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15);
  padding: 6px 0;
  border-radius: 4px;
  top: 0;
  left: 0;
  background-color: #fff;
  min-width: 100px;
  z-index: 10;
  outline: none;
  list-style-type: none;
  cursor: pointer;
}

.br-dropdown-menu {
  padding: 0;
  cursor: pointer;
  transform-origin: 0 0;
}

.br-dropdown-menu-item {
  list-style: none;
  padding: 10px;
  margin: 0;
  border-bottom: 1px solid rgba(255, 255, 255, 0.05);
  cursor: pointer;
  transition: all 0.3s;
}

.br-dropdown-menu-item:last-child {
  border-bottom: 0;
}

.br-dropdown-menu-item:hover {
  opacity: 0.7;
}

.br-dropdown-menu-item:not(.is-disabled):hover {
  background-color: #e6f7ff;
}

.br-dropdown-link {
  cursor: pointer;
}

.br-dropdown-link:hover {
  opacity: 0.7;
}

.br-dropdown-hidden {
  display: none;
}

.slide-up-enter,
.slide-up-appear {
  opacity: 0;
  animation-timing-function: cubic-bezier(0.23, 1, 0.32, 1);
}

.slide-up-enter,
.slide-up-appear,
.slide-up-leave {
  animation-duration: 0.2s;
  animation-fill-mode: both;
  animation-play-state: paused;
}

.slide-up-leave {
  animation-timing-function: cubic-bezier(0.755, 0.05, 0.855, 0.06);
}

.slide-up-enter.slide-up-enter-active,
.slide-up-appear.slide-up-appear-active {
  animation-name: brSlideUpIn;
  animation-play-state: running;
}

.slide-up-leave.slide-up-leave-active {
  animation-name: brSlideUpOut;
  animation-play-state: running;
}

@keyframes brSlideUpIn {
  0% {
    opacity: 0;
    transform-origin: 0% 0%;
    transform: scaleY(0.8);
  }
  100% {
    opacity: 1;
    transform-origin: 0% 0%;
    transform: scaleY(1);
  }
}

@keyframes brSlideUpOut {
  0% {
    opacity: 1;
    transform-origin: 0% 0%;
    transform: scaleY(1);
  }
  100% {
    opacity: 0;
    transform-origin: 0% 0%;
    transform: scaleY(0.8);
  }
}
