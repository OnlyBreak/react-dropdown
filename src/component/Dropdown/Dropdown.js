import React, { PureComponent, createContext } from 'react';
import ReactDOM from 'react-dom';
import ClickOutside from 'react-click-outside';

import './Dropdown.css';

export const DropdownContext = createContext({});

class Dropdown extends PureComponent {
  static defaultProps = {
    trigger: 'click',
    menuAlign: 'bottom-end'
  };

  constructor(props) {
    super(props);

    this.state = {
      visible: false,
      handleClickMenuItem: () => this.handleClickMenuItem(),
      placement: () => this.placement()
    };
  }

  componentDidMount() {
    this.initEvent();
  }

  handleClickMenuItem() {
    if (this.state.visible) {
      this.setState({
        visible: false
      });
    }
  }

  initEvent() {
    const { trigger } = this.props;
    const triggerElm = ReactDOM.findDOMNode(this.defaultElem);

    if (!triggerElm) {
      return;
    }

    if (trigger === 'hover') {
      triggerElm.addEventListener('mouseenter', this.show.bind(this));
      triggerElm.addEventListener('mouseleave', this.hide.bind(this));
    }

    if (trigger === 'click') {
      triggerElm.addEventListener('click', this.toggleMenu.bind(this));
    }
  }

  toggleMenu() {
    if (this.state.visible) {
      this.hide();
    } else {
      this.show();
    }
  }

  show() {
    this.setState({
      visible: true
    });
  }

  hide() {
    this.setState({
      visible: false
    });
  }

  handleClickOutside() {
    if (this.state.visible) {
      this.setState({ visible: false });
    }
  }

  clearDelayTimer() {
    if (this.delayTimer) {
      clearTimeout(this.delayTimer);
      this.delayTimer = null;
    }
  }

  placement() {
    const { menuAlign } = this.props;

    return menuAlign;
  }

  render() {
    const { menu, children } = this.props;

    return (
      <DropdownContext.Provider value={this.state}>
        <div className="br-dropdown">
          {React.cloneElement(children, {
            ref: ref => (this.defaultElem = ref)
          })}
          {React.cloneElement(menu)}
        </div>
      </DropdownContext.Provider>
    );
  }
}

export default ClickOutside(Dropdown);
