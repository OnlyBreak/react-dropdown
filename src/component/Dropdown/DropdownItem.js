import React, { PureComponent } from 'react'
import { DropdownContext } from './Dropdown'

class DropdownItem extends PureComponent {
  handleClick(handleClickMenuItem) {
    const { onCommand } = this.props

    if (handleClickMenuItem) {
      handleClickMenuItem()
    }

    if (onCommand) {
      onCommand()
    }
  }

  render() {
    return (
      <DropdownContext.Consumer>
        {({ handleClickMenuItem }) => (
          <li
            className="br-dropdown-menu-item"
            onClick={() => this.handleClick(handleClickMenuItem)}
          >
            {this.props.children}
          </li>
        )}
      </DropdownContext.Consumer>
    )
  }
}

export default DropdownItem
