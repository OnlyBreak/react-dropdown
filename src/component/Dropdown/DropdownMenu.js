import React, { PureComponent } from 'react';
import Animate from 'rc-animate';
import { DropdownContext } from './Dropdown';
import Popper from 'popper.js';

class DropdownMenu extends PureComponent {
  componentWillUnmount() {
    if (this.popperElem) {
      this.popperElem.destroy();
    }
  }

  getTransitionName() {
    const { placement = '' } = this.props;

    if (placement.indexOf('top') >= 0) {
      return 'slide-down';
    }
    return 'slide-up';
  }

  render() {
    return (
      <DropdownContext.Consumer>
        {({ visible, placement }) => {
          const Box = props => {
            return (
              <div
                style={{ display: props.visible ? 'block' : 'none' }}
                className="br-dropdown-inner"
                ref={ref => this.renderPopper(ref, placement)}
              >
                <ul className="br-dropdown-menu">{this.props.children}</ul>
              </div>
            );
          };

          return (
            <Animate
              component=""
              exclusive
              showProp="visible"
              transitionName={this.getTransitionName()}
            >
              <Box visible={visible} />
            </Animate>
          );
        }}
      </DropdownContext.Consumer>
    );
  }

  renderPopper(ref, placement) {
    if (!ref) {
      return;
    }

    this.popperElem = new Popper(ref.parentNode, ref, {
      placement: placement(),
      modifiers: {
        computeStyle: {
          gpuAcceleration: false
        }
      }
    });
  }
}

export default DropdownMenu;
