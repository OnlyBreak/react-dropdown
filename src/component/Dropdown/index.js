import Dropdown from './Dropdown'
import DropdownMenu from './DropdownMenu'
import DropdownItem from './DropdownItem'

Dropdown.Menu = DropdownMenu
Dropdown.Item = DropdownItem

export default Dropdown
