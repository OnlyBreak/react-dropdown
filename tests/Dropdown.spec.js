import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';

import Dropdown from '../src/component/Dropdown';

describe('Advanced usage', () => {
  const state = {
    menuAlign: 'bottom',
    trigger: 'hover'
  };

  const dropdown = mount(
    <Dropdown
      menu={
        <Dropdown.Menu>
          <Dropdown.Item>1</Dropdown.Item>
          <Dropdown.Item>2</Dropdown.Item>
        </Dropdown.Menu>
      }
      menuAlign={state.menuAlign}
      trigger={state.trigger}
    >
      <span className="dropdown-link">dropdown</span>
    </Dropdown>
  );

  it('allows us to set props', () => {
    expect(dropdown.props().trigger).toEqual('hover');
    dropdown.setProps({ trigger: 'click' });
    expect(dropdown.props().trigger).toEqual('click');
  });

  it('renders correctly', () => {
    dropdown.find('span.dropdown-link').simulate('mouseEnter');
    expect(dropdown.render()).toMatchSnapshot();
  });
});
