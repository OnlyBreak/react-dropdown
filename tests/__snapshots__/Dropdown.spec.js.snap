// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advanced usage renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "br-dropdown",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "dropdown-link",
        },
        "children": Array [
          Object {
            "data": "dropdown",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "br-dropdown-inner",
            "style": "display: none; position: absolute;",
            "x-placement": "bottom",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "br-dropdown-menu",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "br-dropdown-menu-item",
                  },
                  "children": Array [
                    Object {
                      "data": "1",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "class": "br-dropdown-menu-item",
                    },
                    "children": Array [
                      Object {
                        "data": "2",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "li",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                Object {
                  "attribs": Object {
                    "class": "br-dropdown-menu-item",
                  },
                  "children": Array [
                    Object {
                      "data": "2",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "class": "br-dropdown-menu-item",
                    },
                    "children": Array [
                      Object {
                        "data": "1",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "li",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "ul",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
            "x-placement": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
            "x-placement": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "br-dropdown-inner",
          "style": "display: none; position: absolute;",
          "x-placement": "bottom",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "br-dropdown-menu",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "br-dropdown-menu-item",
                },
                "children": Array [
                  Object {
                    "data": "1",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "li",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "class": "br-dropdown-menu-item",
                  },
                  "children": Array [
                    Object {
                      "data": "2",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "class": "br-dropdown-menu-item",
                },
                "children": Array [
                  Object {
                    "data": "2",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "li",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "class": "br-dropdown-menu-item",
                  },
                  "children": Array [
                    Object {
                      "data": "1",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "li",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "ul",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "dropdown-link",
          },
          "children": Array [
            Object {
              "data": "dropdown",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
          "x-placement": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
          "x-placement": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
